### ISSUE TITLE
Question about the true.asm (can you shave one more byte off? :) )

### ISSUE THREAD

**Issue by vaikas** [tianon/dockerfiles#126]
As per:
https://blog.rchapman.org/posts/Linux_System_Call_Table_for_x86_64/

the errorcode used by syscall exit comes from the the %rdi register

Here the intent (I believe at least) is to zero out the %rdi register cheaply:
https://github.com/tianon/dockerfiles/blob/master/true/true.asm#L60

But from reading this, it would seem to me that will zero out the %rdx and not the %rdi:
https://www.felixcloutier.com/x86/cwd:cdq:cqo

If I remove that line, build with nasm, doing docker build & run and inspecting the exit code with docker ps will still return 0. 

```
4c278035ed47   true                     "/true"                  2 seconds ago       Exited (0) 1 second ago                  zen_joliot
```

I also modified the code by moving some other value to rdi, for example, replacing the cdq with this:

```
	mov	rdi, 2
```

and then build with nasm, doing docker build & run and inspecting the exit code with docker ps would return 2 like so:

```
d0dadd05206b   74bba39690e4             "/true"                  42 seconds ago      Exited (2) 41 seconds ago                sharp_jackson
```

So, I'm just curious if I'm misunderstanding something, or if that line could actually be removed and you could save one more byte. Saving the byte is not the goal of mine necessarily, but just trying to understand if I'm misunderstanding something :) My hypothesis is that the elf initialize code sets the %rdi to 0 and that's why it works currently, since the cdq is a nop?
https://elixir.bootlin.com/linux/v3.18/source/arch/x86/include/asm/elf.h#L171

@tiborvass for thoughts since he seems to have made the change. I understand this change was made 7 years ago, but just happened to stumble on this and was just playing around with some stuff and was a bit confused.

**Comment by tiborvass** [comment]
Wow, thank you I think you're right. The error came from the original source from https://web.archive.org/web/20140528193301/http://blog.markloiseau.com/2012/05/tiny-64-bit-elf-executables/ and I guess I didn't catch it when I copied it.

I leave it to @tianon whether he wants to rely on [Linux's behavior that resets the RDI register to 0](https://elixir.bootlin.com/linux/latest/source/arch/x86/include/asm/elf.h#L108), in which case we could indeed just remove `cdq`. Otherwise `xor rdi, rdi` should be okay.

**Comment by tianon** [comment]
I was reading https://www.muppetlabs.com/~breadbox/software/tiny/teensy.html recently, and wondering if it would be worth abusing ELF and going even further :sweat_smile:

There's some risk to doing so, but I hope nobody is actually relying on `tianon/true` for anything serious these days. :joy:

**Comment by vaikas** [comment]
LOL, from the article:
```
Of course, half of the values in this file violate some part of the ELF standard, and it's a wonder that Linux will even consent to sneeze on it, much less give it a process ID. This is not the sort of program to which one would normally be willing to confess authorship.
```

Looks pretty legit :) Maybe at that point consider renaming it to: tianon/true-yolo

Thanks for the pointer!

**Comment by tianon** [comment]
Huge delay later, but great naming suggestion -- implemented `tianon/true:yolo` in #379 :joy:

**Comment by vaikas** [comment]
Oh, that's so awesomesauce :) Really made my day 🤣
